### ISSUE TITLE
Reserved slots to unknown categories in CategoryEncoder

### ISSUE THREAD

**Issue by larvasapiens** [numenta/nupic#3094]
Sometimes you have access to a dataset for training, but are not sure if there will be any new categories in the future. The Adaptative Scalar Encoder is not an ideal option, because it means that the SP and TM will have to re-map the adapted bits, loosing some of the work that has been done. The Random Scalar Encoder does not work well in certain situations, when you need to ensure each category is treated completely separately. I came up with a small update to the Category Encoder. When you create it, you can reserve some slots for potential new categories. When a new category is fed, the encoder maps it to one of the previously reserved slots without affecting any of the other categories. If the encoder runs out of slots, newer categories will be treated as Unknown.

I already made a pull request for this, sorry for not opening the issue before ( #3093 ).

**Comment by rhyolight** [comment]
@numenta/nupic-committers We need to discuss this everyone. See @larvasapiens 's PR above. Any thoughts?

**Comment by ywcui1990** [comment]
I commented on @larvasapiens 's PR. I also pasted my comments below:

"I like the idea to make the CategoryEncoder adaptive. Do we need to really hard code "nAdditionalCategorySlots" upfront? Since SDR has huge capacity, why cannot we dynamically add new categories in run time?"

**Comment by larvasapiens** [comment]
How would be the implementation of that @ywcui1990 ?

**Comment by subutai** [comment]
> Since SDR has huge capacity, why cannot we dynamically add new categories in run time?

We just have to be careful that any downstream code (such as the classifier) can also handle adding new categories dynamically.

**Comment by scottpurdy** [comment]
I believe the desired functionality can be found in the SDRCategoryEncoder:
https://github.com/numenta/nupic/blob/master/src/nupic/encoders/sdrcategory.py

Does this work?

**Comment by cogmission** [comment]
I forgot about this one, that's right... It does work for unknown
categories, doesn't it?

On Wed, Apr 20, 2016 at 1:40 PM, Scott Purdy <notifications@github.com>
wrote:

> I believe the desired functionality can be found in the SDRCategoryEncoder:
>
> https://github.com/numenta/nupic/blob/master/src/nupic/encoders/sdrcategory.py
>
> Does this work?
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/numenta/nupic/issues/3094#issuecomment-212553027>
>



-- 
*With kind regards,*

David Ray
Java Solutions Architect

*Cortical.io <http://cortical.io/>*
Sponsor of:  HTM.java <https://github.com/numenta/htm.java>

d.ray@cortical.io
http://cortical.io

**Comment by larvasapiens** [comment]
@scottpurdy Yeah, almost. After a few quick tests I noted 3 things that differ from my "reserved slots" approach.

1 - Performance: The SDRCategoryEncoder uses random to assign categories. This overhead doesn't exists on the original Category Encoder.
2 - Sparseness: The SDRCategoryEncoder provides a more sparse representation. I don't know if that's actually useful, as one generally passes the output of the Encoder to the SpatialPooler anyway.
3 - Handling of Slots Run Out: The SDRCategoryEncoder gave me a RuntimeError when it couldn't find a unique pattern to represent a novel input. In the original CategoryEncoder, as in my approach, it just returns the representation of the category <UNKOWN>. Which one is more useful depends on the use case, I guess.

**Comment by breznak** [comment]
1/ I don't think speed of Random is a limiting factor that could/should be
taken into account, in comparison with the other encoder methods
2/ It is in real world situations. I remember a discussion (and tests?)
that Scalar/Category encoders are not very useful for real(float) numbers.
(They are kept in codebase for simplicity and demonstration), RDSE (and
derivates) should be used for "generic" use, unless you know what you're
doing.
3/ Strange, as it kinda contradicts what you say in 2/ (>> sparsity).
Didn't you have some odd w/n values?

cheers,

**Comment by larvasapiens** [comment]
After thinking about it, the 3rd scenario is not likely to be a problem, as one should pass a big-enough "n" parameter. If the adventage of the sparseness is worth it like you say, then the SDREncoder should do a better job in most cases :D

**Comment by scottpurdy** [comment]
@larvasapiens - I think there are a number of different scenarios. I personally think that the SDRCategoryEncoder should not store the generated representations and instead use the input value as a seed to the RNG to get deterministic encodings. This would break the ability to "decode" SDRs to categories though so simpler to leave as is for now.

The CategoryEncoder could sometimes be easier to specify the categories at the beginning and sometimes easier to just say how many categories. And the "overflow" bucket sometimes isn't desirable.

So if someone thinks there is a real problem with the existing encoders that can't be solved with a simple mechanism before feeding your values to the encoder then I am open to a proposal. Otherwise it would be good to at least do a documentation update in the code since it looks like it could be improved a lot.
